core/src/main/java/org/elasticsearch/index/translog/TranslogDeletionPolicy.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.translog;
+
+import org.apache.lucene.util.Counter;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public final class TranslogDeletionPolicy {
+
+    /** Records how many views are held against each
+     *  translog generation */
+    protected final Map<Long, Counter> translogRefCounts = new HashMap<>();
+
+    /**
+     * the translog generation that is requires to properly recover from the oldest non deleted
+     * {@link org.apache.lucene.index.IndexCommit}.
+     */
+    private long minTranslogGenerationForRecovery = 1;
+
+    public synchronized void setMinTranslogGenerationForRecovery(long newGen) {
+        if (newGen < minTranslogGenerationForRecovery) {
+            throw new IllegalArgumentException("minTranslogGenerationForRecovery can't go backwards. new [" + newGen + "] current [" +
+                minTranslogGenerationForRecovery+ "]");
+        }
+        minTranslogGenerationForRecovery = newGen;
+    }
+
+    /**
+     * acquires the basis generation for a new view. Any translog generation above, and including, the returned generation
+     * will not be deleted until a corresponding call to {@link #releaseTranslogGenView(long)} is called.
+     */
+    public synchronized long acquireTranslogGenForView() {
+        translogRefCounts.computeIfAbsent(minTranslogGenerationForRecovery, l -> Counter.newCounter(false)).addAndGet(1);
```

Comment:
Too bad Java does not have a way to signal an intentionally unused Lambda parameter like `_` in Scala.